### PR TITLE
Track user clicks on right panel and basic shortcuts

### PR DESCRIPTION
**Pull request by PPiedel**
To run it locally you should follow these steps:
1. 
```bash
$ curl -sSL https://raw.githubusercontent.com/bitnami/bitnami-docker-matomo/master/docker-compose.yml > docker-compose.yml
$ docker-compose up -d
```
2. make changes to port mapping if needed
3. go to matomo site and configure new website (localhost:3000)
4. add 
```
analytics {
  engine: "Matomo"
  url: "http://localhost:<matomo-port>/matomo.php"
  siteId: "<siteId>"
}
```
to your ui/server/develConf/sample/application.conf
5. run
```bash
./buildServer.sh
./runServer.sh
npm start
````
in the appropriate directories

### CODE REVIEW

File: ui/client/containers/EspApp.js [hunk 1]
```diff
@@ -24,17 +24,25 @@ import '../assets/fonts/fonts.less'
...
+    this.generateUniqueUserId()
     this.mountedHistory = this.props.history.listen((location, action) => {
       if (action === "PUSH") {
         this.props.actions.urlChange(location)
       }
     })
   }
 
+  generateUniqueUserId() {
+    if (!localStorage.getItem(userId)) {
```
Comment (lciolecki) [inline ui/client/containers/EspApp.js]: Look at SystemUtils and put evrything connected with localStorage in this class.

**Comment by lciolecki** [comment]
You should also support situation when matomo analytics is disabled.

File: ui/client/components/right-panel/UserRightPanel.js [hunk 2]
```diff
@@ -136,17 +136,47 @@ class UserRightPanel extends Component {
...
-        {name: "save" + (!conf.saveDisabled ? "*" : ""), visible: this.props.capabilities.write, disabled: conf.saveDisabled, onClick: this.save, icon: InlinedSvgs.buttonSave},
-        {name: "migrate", visible: this.props.capabilities.deploy && !_.isEmpty(this.props.featuresSettings.remoteEnvironment), disabled: !conf.deployPossible, onClick: this.migrate, icon: InlinedSvgs.buttonMigrate},
-        {name: "compare", onClick: this.compareVersions, icon: 'compare.svg', disabled: this.hasOneVersion()},
-        {name: "import", visible: this.props.capabilities.write, disabled: false, onClick: this.importProcess, icon: InlinedSvgs.buttonImport, dropzone: true},
-        {name: "JSON", disabled: !this.props.canExport, onClick: this.exportProcess, icon: InlinedSvgs.buttonExport},
-        {name: "PDF", disabled: !this.props.canExport, onClick: this.exportProcessToPdf, icon: InlinedSvgs.pdf},
-        {name: "archive", onClick: this.archiveProcess, disabled: this.isRunning(), icon: 'archive.svg', visible: this.props.capabilities.write}
-      ]
-    },
+        panelName: "Process",
```
[the remark this answers is not in the document]
Comment (PPiedel) [reply ui/client/components/right-panel/UserRightPanel.js]: I've made changes regarding onClick and onDrop methods as well

File: ui/client/containers/EspApp.js (hunk 1 above)
Comment (PPiedel) [reply ui/client/containers/EspApp.js]: Done

File: ui/client/tracking/UserTracker.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
+import {userId} from "../containers/EspApp"
+
+const matomoTracker = () => {
+  const MatomoTracker = require('matomo-tracker')
+  return MatomoTracker(2, "http://localhost:80/matomo.php")
```
[the remark this answers is not in the document]
Comment (PPiedel) [reply ui/client/tracking/UserTracker.js]: Done

File: ui/client/components/right-panel/UserRightPanel.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (PPiedel) [reply ui/client/components/right-panel/UserRightPanel.js]: No, I've made changes regarding onClick and onDrop methods as well

**Comment by PPiedel** [comment]
> Camelcase.

Done

**Comment by PPiedel** [comment]
> We shouldn't use reqiure..

Done

File: ui/client/analytics/AnalyticsMiddleware.js [hunk 4]
```diff
@@ -0,0 +1,18 @@
+import Analytics from "./Analytics";
+
+export const analyticsMiddleware = (store) => next => action => {
+  const state = !_.isEmpty(store) ? store.getState() : null
+  const analyticsSettings = !_.isEmpty(state) ? state.settings.analyticsSettings : null
```
Comment (lciolecki) [inline ui/client/analytics/AnalyticsMiddleware.js]: Simply way: _.get(state, 'settings.analyticsSettings')? :)

File: ui/client/containers/EspApp.js [hunk 5]
```diff
@@ -24,10 +24,12 @@ import '../assets/fonts/fonts.less'
 import '../stylesheets/main.styl'
 import '../app.styl'
 import ErrorHandler from "./ErrorHandler"
+import SystemUtils from "../common/SystemUtils";
 
 class EspApp extends React.Component {
 
   componentDidMount() {
+    SystemUtils.generateUniqueUserId()
```
Comment (lciolecki) [inline ui/client/containers/EspApp.js]: Move it to SystemUtil.getUserId.

File: ui/client/analytics/MatomoAnalytics.js [hunk 6]
```diff
@@ -0,0 +1,20 @@
+import SystemUtils from "../common/SystemUtils";
+import * as MatomoTracker from "matomo-tracker";
+import AnalyticsEngine from "./AnalyticsEngine";
+
+export default class MatomoAnalytics extends AnalyticsEngine {
+
+  constructor(analyticsSettings) {
+    super();
+    this.siteId = analyticsSettings.siteId
+    this.url = analyticsSettings.url
```
Comment (lciolecki) [inline ui/client/analytics/MatomoAnalytics.js]: You cane use _.assign: https://lodash.com/docs/4.17.15#assign.

File: ui/client/analytics/AnalyticsMiddleware.js (hunk 4 above)
Comment (PPiedel) [reply ui/client/analytics/AnalyticsMiddleware.js]: Done

File: ui/client/containers/EspApp.js (hunk 5 above)
Comment (PPiedel) [reply ui/client/containers/EspApp.js]: Done

File: ui/client/analytics/MatomoAnalytics.js (hunk 6 above)
Comment (PPiedel) [reply ui/client/analytics/MatomoAnalytics.js]: It's simple assignment to class fields. Why lodash is better than simple = ?

Comment (lciolecki) [reply ui/client/analytics/MatomoAnalytics.js]: One line :) _.assign(this, _.pick(analyticsSettings, ["siteId", "url"]))

Comment (PPiedel) [reply ui/client/analytics/MatomoAnalytics.js]: Ok, done

File: ui/client/analytics/Analytics.js [hunk 7]
```diff
@@ -0,0 +1,12 @@
+import MatomoAnalytics from "./MatomoAnalytics";
+
+export default class Analytics {
+
+  constructor(analyticsSettings) {
+    if (analyticsSettings.engine === 'Matomo') {
```
Comment (lciolecki) [inline ui/client/analytics/Analytics.js]: This should be declared as const. I mean const ENGINE_MATOMO = 'Matomo'  and next: 
```
if (analyticsSettings.engine === ENGINE_MATOMO)
```

File: ui/client/common/SystemUtils.js [hunk 8]
```diff
@@ -38,6 +40,14 @@ class SystemUtils {
       return config;
     });
   }
+
+  getUserId() {
+    if (!localStorage.getItem(userId)) {
```
Comment (lciolecki) [inline ui/client/common/SystemUtils.js]: Better way will be `this.getItem(userId) !== null`

File: ui/client/analytics/AnalyticsMiddleware.js [hunk 9]
```diff
@@ -0,0 +1,18 @@
+import Analytics from "./Analytics";
```
Comment (lciolecki) [inline ui/client/analytics/AnalyticsMiddleware.js]: Remove all semicolon.. :)

File: ui/client/analytics/Analytics.js (hunk 7 above)
Comment (PPiedel) [reply ui/client/analytics/Analytics.js]: Done

File: ui/client/common/SystemUtils.js (hunk 8 above)
Comment (PPiedel) [reply ui/client/common/SystemUtils.js]: I've changed it to _ .isEmpty... - if the user key will be "" it also not acceptable

File: ui/client/analytics/AnalyticsMiddleware.js (hunk 9 above)
Comment (PPiedel) [reply ui/client/analytics/AnalyticsMiddleware.js]: Done

**Pull request merged**